src/rpc/server.zig

```diff
@@ -0,0 +1,272 @@
+    if (req.body()) |body| {
+        var jrpc_request = std.json.parseFromSlice(t.JsonRpcRequest, req.arena, body, .{}) catch |err| {
+            return respondJsonRpcError(
+                .null,
+                response,
+                t.jrpc_error_code_invalid_params,
+                std.fmt.allocPrint(req.arena, "Invalid request body, expected type JsonRpcRequest: {any}", .{err}) catch unreachable,
+            );
+        };
+
+        defer jrpc_request.deinit();
+        return methodCall(req.arena, ctx.state.rpc_request_processor, &jrpc_request.value, response);
+    } else {
+        // return no body
+        return respondJsonRpcError(
+            .null,
+            response,
+            t.jrpc_error_code_invalid_request,
+            "Missing request body, expected type JsonRpcRequest",
+        );
+    }
+}
+
+fn optionsHandler(ctx: ReqCtx, req: *httpz.Request, response: *httpz.Response) !void {
+    _ = ctx;
+    var allowed_origin = if (req.headers.get("Origin")) |v| v else "*";
+    response.header("access-control-allow-origin", allowed_origin);
+    response.header("access-control-allow-methods", "POST, GET, OPTIONS");
+    response.header("access-control-max-age", "86400");
+    response.header("connection", "Close");
+    response.header("cache-control", "no-cache");
+    response.status = 200;
+    try response.write();
+}
+
+fn getHandler(ctx: ReqCtx, req: *httpz.Request, response: *httpz.Response) !void {
+    _ = ctx;
+    _ = req;
+    response.status = 200;
+    try response.writer().writeAll("Used HTTP Method is not allowed. POST or OPTIONS is required");
+    try response.write();
+}
+
+inline fn methodCall(allocator: std.mem.Allocator, processor: *RpcRequestProcessor, request: *t.JsonRpcRequest, response: *httpz.Response) !void {
+    inline for (@typeInfo(t.RpcServiceImpl(RpcRequestProcessor, t.Result)).Struct.fields) |field| {
+        if (strEquals(request.method, field.name)) {
+            return RpcFunc(@field(RpcRequestProcessor, field.name)).call(allocator, processor, request, response);
+        }
+    }
+
+    return respondJsonRpcError(
+        request.id,
+        response,
+        t.jrpc_error_code_internal_error,
+        "not implemented",
+    );
+}
+
+inline fn strEquals(value: []const u8, other: []const u8) bool {
+    return std.mem.eql(u8, value, other);
+}
+
+pub fn RpcFunc(comptime func: anytype) type {
+    const S = struct {
+        fn call(allocator: std.mem.Allocator, processor: *RpcRequestProcessor, request: *t.JsonRpcRequest, response: *httpz.Response) !void {
+            if (request.params != .array) {
+                return respondJsonRpcError(
+                    request.id,
+                    response,
+                    t.jrpc_error_code_invalid_params,
+                    std.fmt.allocPrint(allocator, "Invalid params, expected type: array, got: {s} ", .{@tagName(request.params)}) catch unreachable,
```

Comment:
since this fcn is already `anyerror` we should probably keep this as a `try` since the alloc can fail?